### ISSUE TITLE
Joblib hangs without crashing

### ISSUE THREAD

**Issue by paulomalvar** [joblib/joblib#125]
Hi guys,

I went through the open and closed issues and I didn't find a mention of the issue that I'm experiencing with joblib.

I'm building a system that:
1) parses in parallel a bunch of rss feeds (using feedparser)
2) scrapes texts in parallel from the parsed rss feeds. For this I basically use urllib2 and goose-extractor.

In both parallel stages I'm experiencing hangs after going through all the instances that the generators pass to the parallelized functions.

Joblib just hangs and stays inactive in memory without crashing. I haven't been able to spot the conditions that trigger this behavior so I can't provide much more detail.

Is any body else experiencing hangs of this sort?

Regards,
Paulo Malvar

**Comment by RasterBurn** [comment]
@pprett did you have a workaround for mixing sklearn joblib with vanilla joblib?

**Comment by pprett** [comment]
@RasterBurn not really except for using the version bundled with sklearn.

**Comment by sciunto** [comment]
I'm able to reproduce on two windows machines with anaconda python 2.7 and joblib installed with pip.

It hangs each time I tried if I set n_jobs=2. For n_jobs=1, I got the result.
On Linux, I do not have this issue.

Hope it helps a little.

**Comment by sciunto** [comment]
To add a more. I noticed that on windows, the assertion l 488 in forking.py from the multiprocessing module fails.

I quote this piece of code.
```
472     if 'main_path' in data:
473         main_path = data['main_path']
474         main_name = os.path.splitext(os.path.basename(main_path))[0]
475         if main_name == '__init__':
476             main_name = os.path.basename(os.path.dirname(main_path))
477 
478         if main_name != 'ipython':
479             import imp
480 
481             if main_path is None:
482                 dirs = None
483             elif os.path.basename(main_path).startswith('__init__.py'):
484                 dirs = [os.path.dirname(os.path.dirname(main_path))]
485             else:
486                 dirs = [os.path.dirname(main_path)]
487 
488             assert main_name not in sys.modules, main_name

```

**Comment by lesteve** [comment]
@sciunto your problem is well-known and isn't relevant to this issue: n_jobs != 1 does not work on Windows in the `__main__ ` namespace. The root cause lies in the multiprocessing module and not joblib per se. See for example [this](http://stackoverflow.com/a/3405479) for more details.

**Comment by sciunto** [comment]
thanks!

**Comment by chengguangnan** [comment]
Had a similar issue when the machine is low on memory. The process hangs.

**Comment by jni** [comment]
Has anyone figured this one out yet? I'm getting this issue repeatedly but unpredictably:

<img width="592" alt="joblib hanging" src="https://cloud.githubusercontent.com/assets/492549/16702850/b8781650-4536-11e6-81a3-85bf743a6a6f.png">

**Comment by paulomalvar** [comment]
@jni one thing that helped me in some cases was to check that I wasn't passing an empty list to the generator that feeds the parallelized function. That might not be the issue in your case, but I just thought of mentioning.

Now, I'm experiencing hangs again, but now when trying to classify in parallel documents using a model trained with Keras+Theano.

Weirdly enough if I retrain the model using a lower dimensionality as output of the first neurons layer this doesn't happen.

At first I thought this might be a memory issue, but I'm experiencing the same issue with a machine that has 256GB of RAM. And monitoring the machine I can see the machine is using only 1% of all the available RAM.

**Comment by jni** [comment]
It would also be good to know whether freezes are related to [this sklearn FAQ entry](http://scikit-learn.org/stable/faq.html#why-do-i-sometime-get-a-crash-freeze-with-n-jobs-1-under-osx-or-linux), even though I'm running on 3.5... I just have no idea what's happening.

@paulomalvar the list is definitely not empty, and this all runs fine if I set n_jobs to 1.

**Comment by paulomalvar** [comment]
@jni try specifying the backend as "threading" not "multiprocessing": https://pythonhosted.org/joblib/parallel.html#using-the-threading-backend

**Comment by jni** [comment]
@paulomalvar I'm parallelising Python functions, so that would defeat the purpose...

**Comment by paulomalvar** [comment]
Got it. It's frustrating when it seems one has to do black magic to bypass obscure issues like this...

**Comment by GaelVaroquaux** [comment]
> @paulomalvar I'm parallelising Python functions, so that would defeat the
> purpose...

Granted, but it might still be useful to narrow down the problem.

**Comment by lesteve** [comment]
> It would also be good to know whether freezes are related to this sklearn FAQ entry, even though I'm running on 3.5... I just have no idea what's happening.

This FAQ entry is slightly misleading unfortunately, 'forkserver' is NOT the default start_method for python 3.4+. From https://pythonhosted.org/joblib/parallel.html#bad-interaction-of-multiprocessing-and-third-party-libraries:

> To avoid this problem joblib.Parallel can be configured to use the 'forkserver' start method on Python 3.4 and later. The start method has to be configured by setting the JOBLIB_START_METHOD environment variable to 'forkserver' instead of the default 'fork' start method. However the user should be aware that using the 'forkserver' prevents joblib.Parallel to call function interactively defined in a shell session.

@jni could you try setting the JOBLIB_START_METHOD and see whether that fixes the problem?

**Comment by jni** [comment]
@lesteve [turns out](https://twitter.com/jnuneziglesias/status/753379051469877248) this instance was my own fault, was passing a list of 1 big job to joblib. But I've definitely encountered this hang (for realz) elsewhere, so I'll keep an eye out and give that a go.

**Comment by rueberger** [comment]
I can reproduce this with the following

```
from scipy.optimize import linear_sum_assignment

```

**Comment by lesteve** [comment]
@rueberger your snippet works fine for me on an Ubuntu box. Are you on OSX?

Have you read this link: https://pythonhosted.org/joblib/parallel.html#bad-interaction-of-multiprocessing-and-third-party-libraries?

**Comment by rueberger** [comment]
Yeah, as mentioned above using forkserver didn't help and caused significant slowdowns. The problem occurs on my Ubuntu 14.04 box.

**Comment by lesteve** [comment]
Hmmm I can not reproduce on an Ubuntu 14.04 box either (Conda install).

How did you install your numpy and scipy and which versions are you using? What about the joblib version?

**Comment by rueberger** [comment]
The problem seems to occur stochastically - you might try increasing the number of iterations in the loop.  Attached find the output of `conda list`

[packages.txt](https://github.com/joblib/joblib/files/394092/packages.txt)

**Comment by lesteve** [comment]
Not sure why this is happening and I am not able to reproduce, could you try with n_jobs=1 to see whether this  you can get an error without multi-processing?

Also I noticed a weird thing in your `conda list` output:
```
numpy                     1.10.4                   py35_1  
numpy                     1.11.0                    <pip>
```

In my experience mixing pip and conda for the same package is not the greatest of idea. Could you try uninstalling numpy and reinstalling it with conda. Something like this, although non minimalistic, should do it:

```
pip uninstall -y numpy
pip uninstall -y numpy # just in case
conda uninstall numpy -y
conda install numpy -y
```

**Comment by ibegleris** [comment]
So any resolution on this? I dont get the problem on my Linux PC but as soon as I send it to a cluster it hangs forever.

**Comment by lesteve** [comment]
If you read the comments above, you'll realise that it's not a single problem with a single solution. Start by reading: https://pythonhosted.org/joblib/parallel.html#bad-interaction-of-multiprocessing-and-third-party-libraries.

Because it only happens on your cluster and not on your Linux PC, there is a very tiny chance that we can reproduce your problem. I would advise you to try to reproduce your problem with a simple snippet and work from there.

**Comment by samuela** [comment]
I'm also seeing an error very similar to this. It occurs when attempting to use `cross_val_score` and `GridSeearchCV` from scikit-learn, but I've basically narrowed it down to this issue. I'm on Mac OS X 10.11.6, Python 2.7.10, and joblib 0.10.2.

**Comment by ogrisel** [comment]
This should be solved in 0.20.0 and master.

Please reopen a new issue with a minimal reproduction case if you still observe hanging.
